CMakeLists.txt

```diff
@@ -0,0 +1,289 @@
+    file(MAKE_DIRECTORY "build")
+    add_custom_command(OUTPUT "bitsandbytes/bitsandbytes.metallib"
+                COMMAND xcrun metal -c -o "build/bitsandbytes.air" ${METAL_FILES}
+                COMMAND xcrun metallib "build/bitsandbytes.air" -o "bitsandbytes/bitsandbytes.metallib"
+                DEPENDS "${METAL_FILES}"
+                COMMENT "Compiling Metal kernels"
+                VERBATIM)
+    add_custom_target(metallib DEPENDS "bitsandbytes/bitsandbytes.metallib")
+else()
+    string(APPEND BNB_OUTPUT_NAME "_cpu")
+    set(GPU_SOURCES)
+endif()
+
+
+if(WIN32)
+    # Export all symbols
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+endif()
+
+# Weird MSVC hacks
+if(MSVC)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /arch:AVX2 /fp:fast")
+endif()
+
+set_source_files_properties(${CPP_FILES} PROPERTIES LANGUAGE CXX)
+add_library(bitsandbytes SHARED ${SRC_FILES})
+target_compile_features(bitsandbytes PUBLIC cxx_std_14)
+target_include_directories(bitsandbytes PUBLIC csrc include)
+
+
+if(BUILD_CUDA)
+    target_include_directories(bitsandbytes PUBLIC ${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
+    target_link_libraries(bitsandbytes PUBLIC CUDA::cudart CUDA::cublas CUDA::cusparse)
+    if(NO_CUBLASLT)
+        target_compile_definitions(bitsandbytes PUBLIC NO_CUBLASLT)
+    else()
+        target_link_libraries(bitsandbytes PUBLIC CUDA::cublasLt)
+    endif()
+
+    set_target_properties(bitsandbytes
+        PROPERTIES
+            CUDA_SEPARABLE_COMPILATION ON
+    )
+endif()
+if(BUILD_HIP)
+    list(APPEND CMAKE_PREFIX_PATH /opt/rocm)
+    macro(find_package_and_print_version PACKAGE_NAME)
+      find_package("${PACKAGE_NAME}" ${ARGN})
+      message("${PACKAGE_NAME} VERSION: ${${PACKAGE_NAME}_VERSION}")
+    endmacro()
+    find_package_and_print_version(hipblas REQUIRED)
+    find_package_and_print_version(hiprand REQUIRED)
+    find_package_and_print_version(hipsparse REQUIRED)
+
+    ## hacky way of excluding hip::amdhip64 (with it linked many tests unexpectedly fail e.g. adam8bit because of inaccuracies)
+    set_target_properties(hip::host PROPERTIES INTERFACE_LINK_LIBRARIES "")
+    set_target_properties(hip-lang::host PROPERTIES INTERFACE_LINK_LIBRARIES "")
+    set(CMAKE_HIP_IMPLICIT_LINK_LIBRARIES "")
+
+    target_include_directories(bitsandbytes PRIVATE ${CMAKE_SOURCE_DIR} ${CMAKE_SOURCE_DIR}/include /opt/rocm/include /include)
+    target_link_directories(bitsandbytes PRIVATE /opt/rocm/lib /lib)
+    target_link_libraries(bitsandbytes PUBLIC roc::hipblas hip::hiprand roc::hipsparse)
+
+    target_compile_definitions(bitsandbytes PUBLIC BNB_USE_HIP)
+    set_source_files_properties(${HIP_FILES} PROPERTIES LANGUAGE HIP)
+    set_target_properties(bitsandbytes PROPERTIES LINKER_LANGUAGE CXX)
+
+    if(NO_CUBLASLT)
+	target_compile_definitions(bitsandbytes PUBLIC NO_HIPBLASLT)
+    else()
+	find_package(hipblaslt)
+        target_link_libraries(bitsandbytes PUBLIC roc::hipblaslt)
+    endif()
```

Comment:
Would be preferable if we just always link hipblaslt if possible.
I see we already have `supports_igemmlt()` always return `True` if `torch.version.hip`.

Reply:
Sure, I can change that. Added that part to support NO_CUBLASLT option, but haven't implemented it yet. I will take it out to avoid confusion.

Reply:
RDNA 2 / (3) and CDNA 1 don't support hipblasLt though, only very recent devices. Cuda has the option as well even support is far broader so it should absolutely stay. Python part will need adjustment on which shared library to load, we could probably take a look at how Cuda does it, but simplest solution will be to just check 

```python
if gfx_arch[0] == 9 and gfx_arch[1] in ["0a", "4x"] #no idea if second part of tuple will be exactly need can't test, for me gfx 1030 is (10, 3)
```

and only enable hibblaslt / igemm for that.

Reply:
> Is this related? [ROCm/hipBLASLt#516](https://github.com/ROCm/hipBLASLt/issues/516)

I can use latest torch just fine on gfx1030, so I guess even if we linked it, it would work. But no idea about how it works internally.
> Does this absolutely need to be implemented with hipBLASLt or could another library like rocBLAS be used?

wdym? blasLt is for matrix cores, yes functions could be emulated with rocBLAS, but bitsandbytes works just fine without it. I can use optimizers and 4bit stuff and I'm pretty sure there would be no speed benefit emulating blasLt (if there was we could just improve current fallback code).

> If not, can the build/distribution still be simplified, and link to this library, and then decide whether to call it at runtime based on device support? See [TimDettmers#1103](https://github.com/TimDettmers/bitsandbytes/pull/1103)

I'm biased here. hipblasLt is a very recent library, it's the reason I named my fork 5.6 since it required headers (this repo moved defines though that you wouldn't even need headers which would be nice for pre 6.0 support). Also e.g. Arch still ships [hipblaslt](https://archlinux.org/packages/extra/x86_64/hipblaslt/) separately, don't know about other distros.

If the unified build gets merged for Cuda we can still change it for ROCm later. Imo ROCm should always mirror CUDA unless it's not possible.

Reply:
That's fine. I think mainly the part that got me was `supports_igemmlt()` always returning `True`. I mentioned rocBLAS because my understanding was that it would use the matrix cores / MFMA instructions where supported - or at the very least, [on gfx908](https://github.com/ROCm/rocBLAS/blob/develop/docs/how-to/what-is-rocblas.rst#mi100-gfx908-considerations).

In general my understanding on the CUDA side is that the int8 matmul without tensor cores is going to be deprecated in the future too, but we'll want to fail in a friendly way, so some kind of test at runtime before calling igemmlt would be ideal. IMO not super critical to have a fallback.